### PR TITLE
Add goldendict

### PR DESCRIPTION
**Pull request by proletarius101**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [ ] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** https://github.com/goldendict/goldendict/issues/1285#issuecomment-782814452
- [ ] I own the domain used in the application ID or the domain has a policy for delegating subdomains (e.g. GitHub, SourceForge).
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://github.com/flathub/flathub/wiki/App-Requirements
[maint]: https://github.com/flathub/flathub/wiki/App-Maintenance
[submission]: https://github.com/flathub/flathub/wiki/App-Submission

## TODO
- [x] The `<release>` field

### CODE REVIEW

**Comment by LeandroStanger** [comment]
bot, build org.goldendict.desktop

You can trigger the builds as well in this PR

File: org.goldendict.desktop.yml [hunk 1]
```diff
@@ -0,0 +1,130 @@
+app-id: org.goldendict.desktop
+runtime: org.kde.Platform
+runtime-version: '5.15'
+branch: stable
```
Comment (bilelmoussaoui) [inline org.goldendict.desktop.yml:4]:
```suggestion
```
it's the default

File: org.goldendict.desktop.yml [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        CXX: 'ccache g++'
+        QMAKEPATH: '/app/lib'
+        LD_LIBRARY_PATH: ${LD_LIBRARY_PATH}:/app/lib/:/usr/lib
+    config-opts:
+      - DEFINES+=USE_DBUS
+      - INCLUDEPATH+="${FLATPAK_DEST}/include/QtWebKitWidgets"
+      - INCLUDEPATH+="${FLATPAK_DEST}/include/QtWebKit"
+      - CONFIG+=release
+    build-commands:
+      - sed -i "s,Icon=goldendict,Icon=${FLATPAK_ID}," redist/goldendict.desktop
```
Comment (bilelmoussaoui) [inline org.goldendict.desktop.yml:116]: Use desktop-file-edit

File: org.goldendict.desktop.yml [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      - sed -i "s,<launchable type="desktop-id">goldendict.desktop</launchable>,<launchable type="desktop-id">${FLATPAK_ID}.desktop</launchable>," redist/goldendict.appdata.xml
```
Comment (bilelmoussaoui) [inline org.goldendict.desktop.yml:117]: Do that as a separate patch & send it upstream if please.

File: org.goldendict.desktop.yml [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      - make -j$(nproc)
+      - make -j$(nproc) install
```
Comment (bilelmoussaoui) [inline org.goldendict.desktop.yml:119]:
```suggestion
      - make -j $FLATPAK_BUILDER_N_JOBS
      - make -j $FLATPAK_BUILDER_N_JOBS install
```

File: org.goldendict.desktop.yml (hunk 2 above)
Comment (proletarius101) [reply org.goldendict.desktop.yml:116]: Fixed

File: org.goldendict.desktop.yml (hunk 3 above)
Comment (proletarius101) [reply org.goldendict.desktop.yml:117]: That just renames `goldendict` to its flatpak id. Obviously won't be in the upstream. Using three (patch, desktop-file-edit, appstream-util) looks a little messy. But of course that could be fixed in a better way if  appstream-util could edit the metadata more eligantly

File: org.goldendict.desktop.yml (hunk 4 above)
Comment (proletarius101) [reply org.goldendict.desktop.yml:119]: Fixed

**Comment by proletarius101** [comment]
bot, build org.goldendict.desktop

**Comment by proletarius101** [comment]
bot, build org.goldendict.desktop

File: org.goldendict.desktop.yml [hunk 5]
```diff
@@ -113,10 +112,13 @@ modules:
...
       - CONFIG+=release
     build-commands:
-      - sed -i "s,Icon=goldendict,Icon=${FLATPAK_ID}," redist/goldendict.desktop
-      - sed -i "s,<launchable type="desktop-id">goldendict.desktop</launchable>,<launchable type="desktop-id">${FLATPAK_ID}.desktop</launchable>," redist/goldendict.appdata.xml
-      - make -j$(nproc)
-      - make -j$(nproc) install
+      - desktop-file-edit --set-icon=${FLATPAK_ID} redist/goldendict.desktop
+      # Put the version to the metadata
+      - appstream-util modify redist/goldendict.appdata.xml releases "<release version=\"`cat version.txt`\" date=\"`date --iso-8601`\"/>"
+      - sed -i -e "s,<releases>&lt;,<releases><," -e "s,&gt;</releases>,></releases>," redist/goldendict.appdata.xml
```
Comment (bilelmoussaoui) [inline org.goldendict.desktop.yml:127]: why not do that as part of your patch?

File: org.goldendict.desktop.yml (hunk 3 above)
Comment (bilelmoussaoui) [reply org.goldendict.desktop.yml]: Why it won't be in the upstream?

File: org.goldendict.desktop.yml (hunk 5 above)
Comment (proletarius101) [reply org.goldendict.desktop.yml:127]: Because it varies from version to version?

Comment (bilelmoussaoui) [reply org.goldendict.desktop.yml:127]: That's why it should be contributed upstream.

File: org.goldendict.desktop.yml (hunk 3 above)
Comment (proletarius101) [reply org.goldendict.desktop.yml]: Because usually the `Icon=` is like `goldendict` rather than `org.goldendict.desktop`? I can't find a good reason to persuade the upstream. They would say like "it's the flatpak-only format".

Comment (bilelmoussaoui) [reply org.goldendict.desktop.yml]: It's not the Flatpak-only format. see https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-id-generic & https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#file-naming.

Comment (proletarius101) [reply org.goldendict.desktop.yml]: Great to know. That's a good backup since the specs say so. I'll have a try. But could it be just a link to the PR before the upstream accepts it? The PR may have any destinies anyway.

File: org.goldendict.desktop.yml (hunk 5 above)
Comment (proletarius101) [reply org.goldendict.desktop.yml:127]: I realized they write the versions in the build process: https://github.com/goldendict/goldendict/blob/master/.travis.yml
````
  - export VERSION=$(git rev-parse --short HEAD) # linuxdeployqt uses this for naming the file
  - ./linuxdeployqt-continuous-x86_64.AppImage appdir/usr/share/applications/*.desktop -appimage
````
And it just not compatible with hard-code the version in the metadata.xml. So I don't think they will accept this.

**Comment by proletarius101** [comment]
bot, build org.goldendict.desktop

**Comment by proletarius101** [comment]
bot, build org.goldendict.desktop

**Comment by proletarius101** [comment]
bot, build org.goldendict.desktop

**Comment by proletarius101** [comment]
@bilelmoussaoui 
As per https://salsa.debian.org/debian/eb/-/merge_requests/3#note_228998 they use `dh_update_autotools_config` to update `config.guess` and `config.sub` which sounds like a [common practice](https://man7.org/linux/man-pages/man1/dh_update_autotools_config.1.html). Maybe I should just download the two files at every build as what I'm doing, since there is no `dh_update_autotools_config` in the SDK?

File: org.goldendict.desktop.yml (hunk 5 above)
Comment (nedrichards) [reply org.goldendict.desktop.yml]: That's why the appdata should go upstream and have the version written out as part of the build process as well.

Comment (proletarius101) [reply org.goldendict.desktop.yml]: > That's why the appdata should go upstream and have the version written out as part of the build process as well.

Didn't get it. The version is written in the build process, since they use `linuxdeployqt`. How can I even persuade them to change the build tool?

> have the version written out as part of the build process as well.

The appdata with version is only bundled in the appimage. The only way to "solve" that is to contribute to `linuxdeployqt`??

File: org.goldendict.desktop.yml [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      - /share/doc
+    sources:
+      - type: git
+        url: https://salsa.debian.org/debian/eb.git
+        tag: debian/4.4.3-13
+        x-checker-data:
+          type: anitya
+          project-id: 155542
+          tag-template: debian/$version
+      # Update config.guess & config.sub to allow builds for aarch64 and others
```
Comment (barthalion) [inline org.goldendict.desktop.yml:69]: Just do something like this: https://github.com/flathub/org.glimpse_editor.Glimpse/blob/cd50b8421b2719438aa2b0518a15b1254a16c5a0/org.glimpse_editor.Glimpse.json#L52

File: org.goldendict.desktop.yml [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+          type: anitya
+          project-id: 7629
+          url-template: 'https://downloads.xiph.org/releases/ao/libao-$version.tar.gz'
+
+  - name: goldendict
+    buildsystem: qmake
+    build-options:
+      env:
+        CC: 'ccache gcc'
+        CXX: 'ccache g++'
```
Comment (barthalion) [inline org.goldendict.desktop.yml:116]: What's the point here?

File: org.goldendict.desktop.yml [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    build-commands:
+      # https://github.com/goldendict/goldendict/pull/1354
+      - desktop-file-edit --set-icon=${FLATPAK_ID} redist/goldendict.desktop
+      # Put the version to the metadata
+      - export VERSION=$(git rev-parse --short HEAD) # The same version as the upstream
+      - appstream-util modify redist/goldendict.appdata.xml releases "<release version=\"`cat version.txt` (${VERSION}) \" date=\"`date --iso-8601`\"/>"
+      - sed -i -e "s,<releases>&lt;,<releases><," -e "s,&gt;</releases>,></releases>," redist/goldendict.appdata.xml
+
+      - make -j$FLATPAK_BUILDER_N_JOBS
+      - make -j$FLATPAK_BUILDER_N_JOBS install
```
Comment (barthalion) [inline org.goldendict.desktop.yml:133]: Doesn't it mean you run make and make install twice here? `qmake` buildsystem should already do that

File: org.goldendict.desktop.yml (hunk 6 above)
Comment (proletarius101) [reply org.goldendict.desktop.yml:69]: That's elegant

File: org.goldendict.desktop.yml (hunk 7 above)
Comment (proletarius101) [reply org.goldendict.desktop.yml:116]: To build faster locally? Or that's not used since it's qmake?

Comment (proletarius101) [reply org.goldendict.desktop.yml:116]: And yes it doesn't work. Removed.

File: org.goldendict.desktop.yml (hunk 8 above)
Comment (proletarius101) [reply org.goldendict.desktop.yml:133]: Wow, great to know.

**Comment by proletarius101** [comment]
bot, build org.goldendict.desktop

**Comment by barthalion** [comment]
@proletarius101 Are files here in sync with the work done upstream?

**Comment by proletarius101** [comment]
> @proletarius101 Are files here in sync with the work done upstream?

No. In my todo list

**Comment by proletarius101** [comment]
bot, build org.goldendict.GoldenDict

**Comment by proletarius101** [comment]
Hmm, so why did it strip the `.desktop` as `-exec`

````
/usr/bin/qmake -install qinstall -exe /run/build/goldendict/redist/org.goldendict.GoldenDict.desktop /app/share/applications/org.goldendict.GoldenDict.desktop
strip /app/share/applications/org.goldendict.GoldenDict.desktop
strip: /app/share/applications/org.goldendict.GoldenDict.desktop: file format not recognized
````

**Comment by proletarius101** [comment]
bot, build org.goldendict.GoldenDict

**Comment by proletarius101** [comment]
Another problem is goldendict hasn't been tagged since 2016: https://github.com/goldendict/goldendict/releases. Is it compulsory to use the tagged version? @barthalion

**Comment by proletarius101** [comment]
bot, build org.goldendict.GoldenDict
